### PR TITLE
Support qwen2 vl model

### PR DESCRIPTION
**Pull request by yizhang2077**
## Motivation
This PR adding support for Qwen2-VL model, which is also supported by vllm [(here)](https://github.com/vllm-project/vllm/pull/7905) and Imdeploy [(here)](https://github.com/InternLM/lmdeploy/pull/2449)

## Modifications

1. add conversation template and chat template for Qwen2-vl，which is refereced by [here](https://huggingface.co/docs/transformers/main/model_doc/qwen2_vl#usage-example)
2. add image processor for Qwen2VL since the output of processor are different (for example, Qwen2VL image processor will output image_grid_thws as result), and add member **image_grid_thws** into **ImageInputs** 
3. compute **mrope positions** for each request by using **MRotaryEmbedding** in vllm, and record result in **InputMeta** . QWen2VL model will use them as real positions
4. copy qwen2_vl.py from vllm and make some adaptions, make some modifications in **pad_input_ids** function, which will replace image_token_id in input_id by unique image_hash 
5. add test about qwen2vl, which is largely copied from test_vision_openai_server.py

## Checklist

- [x] Format your code according to the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [x] Add unit tests as outlined in the [Contributor Guide](https://github.com/sgl-project/sglang/blob/main/docs/en/contributor_guide.md).
- [ ] Update documentation as needed, including docstrings or example 

## Others
- [ ] There is a bug in transformers about Qwen2VLConfig [(here)](https://github.com/huggingface/transformers/issues/33401), vllm fix it in [(here)](https://github.com/vllm-project/vllm/pull/8837/files#diff-5df65ae2e3f73462207bd2e9a093b7057def48bdb7e5b32736da56af1aa891f1) while it is still not released to pip. SGLang has dependency about vllm. As a result, if we do not use latest vllm or make some modification for transformers, we can not run QWen2VL model correctly.
- [x] mrope relies on **MRotaryEmbedding** in vllm, which is not supported in vllm 0.5.5, so maybe we need update vllm version

## Notice
- this PR will update dependencies about **vllm** (0.5.5->0.6.3) and **transformers**(4.44->4.45.2)

### CODE REVIEW

File: python/sglang/srt/conversation.py
Comment (merrymercy) [inline python/sglang/srt/conversation.py]: can we store the image token in the chat template?

File: python/sglang/srt/model_executor/forward_batch_info.py
Comment (merrymercy) [inline python/sglang/srt/model_executor/forward_batch_info.py]: It seems this class is python only without c++ custom ops. Can you directly copy this class under `srt/layers/rotary_embedding.py`?

File: test/srt/test_qwen2_vl_openai_server.py
Comment (merrymercy) [inline test/srt/test_qwen2_vl_openai_server.py]: Good test! Instead of copying this over. Is it possible to reuse code in the existing `test_vision_openai_server.py`? A few methods:
1. Write a for loop in `test_vision_openai_server.py` to loop over multiple models.
2. Let `TestQWen2VLServer` be a child class of `TestOpenAIVisionServer`

**Review by merrymercy — COMMENTED** [review]
Thanks for the contributions. I left a few comments.

We also did some refactoring recently (https://github.com/sgl-project/sglang/pull/1541, https://github.com/sgl-project/sglang/pull/1538). Could you rebase?

**Comment by yizhang2077** [comment]
> Thanks for the contributions. I left a few comments.
> 
> We also did some refactoring recently (#1541, #1538). Could you rebase?

Sorry for the late reply, I am on vacation abroad these days. Of course I will resolve these conversations and rebase code.

**Comment by yizhang2077** [comment]
> Thanks for the contributions. I left a few comments.
> 
> We also did some refactoring recently (#1541, #1538). Could you rebase?

OK, I have rebased code into lastet version

File: test/srt/test_qwen2_vl_openai_server.py
Comment (yizhang2077) [reply test/srt/test_qwen2_vl_openai_server.py]: done

File: python/sglang/srt/conversation.py
Comment (yizhang2077) [reply python/sglang/srt/conversation.py]: done

File: python/sglang/srt/model_executor/forward_batch_info.py
Comment (yizhang2077) [reply python/sglang/srt/model_executor/forward_batch_info.py]: done

File: test/srt/test_vision_openai_server.py
Comment (merrymercy) [inline test/srt/test_vision_openai_server.py]: Are both of them not supported? What is the reason that they are not supported?

If supporting them is hard, we can add some model-specific adjustment code here to set them automatically, so users do not need to manually set them.

https://github.com/sgl-project/sglang/blob/bbd72bfc8609d1c5d8bc9ebb29b9c3b9e218bb90/python/sglang/srt/model_executor/model_runner.py#L94-L107

**Comment by merrymercy** [comment]
Can this run correctly now without the modification/update of vllm? If so, we can remove "WIP" in the PR title and merge this soon!

**Comment by merrymercy** [comment]
In https://github.com/sgl-project/sglang/pull/1632, I merged a small change from your PR to make you a contributor of this project. This allows your future commits to automatically trigger the CI.

**Comment by yizhang2077** [comment]
> Can this run correctly now without the modification/update of vllm? If so, we can remove "WIP" in the PR title and merge this soon!

I think not, there are still some dependencies for vllm latest version when loading vllmModelConfig
https://github.com/sgl-project/sglang/blob/56503d9bc93a13db22b9314a93c83a4f2ce4c362/python/sglang/srt/model_executor/model_runner.py#L233-L242

File: python/sglang/srt/models/qwen2_vl.py
Comment (Titan-p) [inline python/sglang/srt/models/qwen2_vl.py:607]: I tested the latest version of vllm, but the mrope_positions is set to None. I'm not sure what the issue is.

Comment (yizhang2077) [reply python/sglang/srt/models/qwen2_vl.py:607]: Thanks for testing my PR, cuda graph and radix cache are still not adapted. Maybe you need disable them? (by adding "--disable-radix-cache" and "--disable-cuda-graph")

File: test/srt/test_vision_openai_server.py
Comment (yizhang2077) [reply test/srt/test_vision_openai_server.py]: Yes, both of them are still not supported currently.  I will try to make adaptions for them this weekend.
1. The reason of disabling radix cache is vllm is still disabling prefix caching and this model code is mainly moved from vllm. But I think this is not hard to solve by using **image_offsets** and I will try it. 
2. The reason of disabling cuda graph is I don't know how to calculate **mrope_positions** in cuda graph capture phase right now. If using cuda graph now, **mrope_positions** is None which will cause error.

File: python/sglang/srt/models/qwen2_vl.py
Comment (ispobock) [inline python/sglang/srt/models/qwen2_vl.py]: Maybe we can modify and use the prefill triton attention kernel here, instead of involving new dependencies.
Add casual option in: https://github.com/sgl-project/sglang/blob/48761171716302446a95c8d9d1fe1a469f12309e/python/sglang/srt/layers/attention/triton_ops/prefill_attention.py#L149

**Comment by ispobock** [comment]
The CI ut issue is caused by the old version of transformers. We need to upgrade transformers to 4.45.2.

**Comment by merrymercy** [comment]
@ispobock We can update the transformers version in the CI https://github.com/sgl-project/sglang/blob/869f1c02c4a7140c674ea92127a45eac0211bf74/.github/workflows/pr-test.yml#L32. There are multiple places.

File: test/srt/test_vision_openai_server.py
Comment (yizhang2077) [reply test/srt/test_vision_openai_server.py]: It support open radix cache now and set disable-cuda-graph to True for qwen2-vl by default. cuda graph can be supported in another PR

File: python/pyproject.toml
Comment (zhyncs) [inline python/pyproject.toml]: I see that you have updated the version of vLLM here. Is there anything blocking this PR at the moment?

**Comment by zhyncs** [comment]
BTW conflicts need to be resolved.

File: python/pyproject.toml
Comment (yizhang2077) [reply python/pyproject.toml]: there are some modification for VllmModelConfig in latest version to adapt qwen2-vl, if not updating vllm, it will cause error when loading qwen2-vl, but it seems VllmModelConfig is not easy to move from vllm
https://github.com/sgl-project/sglang/blob/56503d9bc93a13db22b9314a93c83a4f2ce4c362/python/sglang/srt/model_executor/model_runner.py#L233-L242

File: python/sglang/srt/conversation.py
Comment (merrymercy) [inline python/sglang/srt/conversation.py]: there is already an image_token

File: .github/workflows/pr-test.yml
Comment (merrymercy) [inline .github/workflows/pr-test.yml]: Can you make this a separate PR?

File: python/sglang/srt/layers/attention/triton_ops/prefill_attention.py
Comment (merrymercy) [inline python/sglang/srt/layers/attention/triton_ops/prefill_attention.py]: use two branches instead of dong computation twice for IS_CAUSAL

File: test/srt/test_vision_openai_server.py
Comment (merrymercy) [inline test/srt/test_vision_openai_server.py:295]: Does it also work correctly for video/multi image? If so, it is awesome!

**Review by merrymercy — CHANGES_REQUESTED** [review]

File: test/srt/test_vision_openai_server.py
Comment (yizhang2077) [reply test/srt/test_vision_openai_server.py:295]: Yes, It can work correctly in video/multi image test

File: python/sglang/srt/conversation.py
Comment (yizhang2077) [reply python/sglang/srt/conversation.py]: done

File: .github/workflows/pr-test.yml
Comment (ispobock) [reply .github/workflows/pr-test.yml]: I saw transformers is updated. We also need a PR to upgrade vllm to 0.6.3.

File: python/sglang/srt/layers/attention/triton_ops/prefill_attention.py
Comment (ispobock) [reply python/sglang/srt/layers/attention/triton_ops/prefill_attention.py]: Sure.

File: .github/workflows/pr-test.yml
Comment (merrymercy) [reply .github/workflows/pr-test.yml]: sounds good

File: python/sglang/srt/models/qwen2_vl.py
Comment (zhyncs) [inline python/sglang/srt/models/qwen2_vl.py:33]: remove CacheConfig https://github.com/sgl-project/sglang/pull/1658

File: python/sglang/srt/models/qwen2_vl.py
Comment (zhyncs) [inline python/sglang/srt/models/qwen2_vl.py:36]: use the `init_logger` from SGLang

File: python/sglang/srt/models/qwen2_vl.py
Comment (zhyncs) [inline python/sglang/srt/models/qwen2_vl.py:37]: TODO: It's not difficult to implement in FlashInfer. ref https://github.com/vllm-project/vllm/blob/81ede99ca44a5b3518932a07ea4a76a719e7416e/csrc/cpu/activation.cpp#L62-L67
It can be implemented in subsequent PR.

**Comment by zhyncs** [comment]
BTW remember to run nightly eval after upgrade vllm https://github.com/sgl-project/sglang/actions/workflows/nightly-eval.yml cc @ispobock

**Comment by zhyncs** [comment]
In order to run nightly eval before merging into main, I changed the base branch to qwen2vl. Once you resolve some of the minor issues mentioned above and fix the conflicts with main, we can consider merging and then conduct some compatibility tests.

File: python/pyproject.toml
Comment (zhyncs) [inline python/pyproject.toml]:
```suggestion
srt = ["sglang[runtime_common]", "torch", "vllm==0.6.3.post1"]
```

**Comment by zhyncs** [comment]
https://github.com/sgl-project/sglang/actions/runs/11395749589
This evaluation failed. Could you help fix it and submit another separate PR?

**Comment by merrymercy** [comment]
It seems this PR is merged to [yizhang2077:support-qwen2-vl](https://github.com/yizhang2077/sglang/tree/support-qwen2-vl) by accident? Should we open a new one?

**Comment by yizhang2077** [comment]
> It seems this PR is merged to [yizhang2077:support-qwen2-vl](https://github.com/yizhang2077/sglang/tree/support-qwen2-vl) by accident? Should we open a new one?

It seems this PR is merge into qwen2vl branch，and when this PR #1711 has merged into main, this PR can merge into main

**Pull request merged**
